include/util/json_renderer.hpp

```diff
@@ -56,7 +56,10 @@ template <typename Out> struct Renderer
         size_t decimalpos = std::find(buffer.begin(), buffer.end(), '.') - buffer.begin();
         if (buffer.size() > (decimalpos + 10))
         {
-            buffer.resize(decimalpos + 10);
+            buffer.clear();
+            fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{0:.10f}"), number.value);
```

Comment:
Why don't just straightway format with {0:.10f} a couple of lines above?

Reply:
Because it is not necessary to preserve the precision of Number type numbers without a decimal point, for example, 116 will become 116.0000000000 after 10 decimal places are reserved

Reply:
Response will like this:
<img width="889" alt="image" src="https://user-images.githubusercontent.com/44461064/215405166-f0254d10-ef25-4e3d-bc28-9f5d55d821a6.png">

Reply:
I think initial problem is caused by the fact that fmt uses so called scientific notation for such small numbers. Can we just try to turn it off? This fix is better than no fix, but it seems we can make it slightly better.

Something which I found after 1 min in Google: https://www.reddit.com/r/cpp_questions/comments/n61o3w/how_to_force_fmt_not_to_use_exponent_notation/?utm_source=share&utm_medium=ios_app&utm_name=iossmf

Reply:
Probably we can just use {:10f} and always do resize to cut trailing zeros - should work well I think.

Reply:
Or it seems that {:f} should force fmt to not use scientific notation - in this case we can just change format string in initial fmt::format call

Reply:
I think you want use {:.10f}, {:10f} means 10 width, but {:.10f} do not cut trailing zeros too.

Reply:
> I think initial problem is caused by the fact that fmt uses so called scientific notation for such small numbers. Can we just try to turn it off? This fix is better than no fix, but it seems we can make it slightly better.
> 
> Something which I found after 1 min in Google: https://www.reddit.com/r/cpp_questions/comments/n61o3w/how_to_force_fmt_not_to_use_exponent_notation/?utm_source=share&utm_medium=ios_app&utm_name=iossmf

I'm also thinking about whether there is a more elegant way, unfortunately I haven't found a better solution yet

Reply:
Besides that let's also add some simple test to avoid regressions in the future. I think this one should work in this case:
```
BOOST_AUTO_TEST_CASE(test_json_issue_6531) {
    std::string output;
    osrm::util::json::Renderer<std::string> renderer(output);
    renderer(0.0000000000017114087924596788);
    BOOST_CHECK_EQUAL(output, "0.0");
}
```

Reply:
This test is very useful, I just checked [this](https://github.com/therion/therion/blob/devel/thdouble.h ) avoid zero, looks like it would be better to directly reformatter for fractional parts larger than 10 digits.
<img width="1020" alt="image" src="https://user-images.githubusercontent.com/44461064/215417090-e4bf9541-cce8-41ad-a0f0-60787a9f42fb.png">

Reply:
I eventually came up with something like this:

```
    void operator()(const Number &number)
    {
        // we don't want to print NaN or Infinity
        BOOST_ASSERT(std::isfinite(number.value));
        // `fmt::memory_buffer` stores first 500 bytes in the object itself(i.e. on stack in this
        // case) and then grows using heap if needed
        fmt::memory_buffer buffer;
        fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:f}"), number.value);

        // find out how many trailing zeros we have
        size_t zeros = 0;
        for (size_t index = buffer.size() - 1; index > 1; --index)
        {
            if (buffer[index] != '0' || buffer[index - 1] == '.') {
                break;
            }
            ++zeros;
        }
        // find position of the decimal point
        size_t pointpos = std::find(buffer.begin(), buffer.end(), '.') - buffer.begin();
        // we are removing trailing zeros and truncating the number to 10 digits after the decimal
        buffer.resize(std::min(buffer.size() - zeros, pointpos + 10));

        write(buffer.data(), buffer.size());
    }
 ```
 
 Not sure there are no bugs so - will return to it later(or feel free to continue with it please :) )

Reply:
Nice, but `fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:f}"), number.value);` only keep 6 decimal placesill.
I came up with like this 
```
void operator()(const Number &number)
  {
      // `fmt::memory_buffer` stores first 500 bytes in the object itself(i.e. on stack in this
      // case) and then grows using heap if needed
      fmt::memory_buffer buffer, tmp_buffer;

      fmt::format_to(std::back_inserter(tmp_buffer), FMT_COMPILE("{0:.10f}"), number.value);
      fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{}"), std::stod(to_string(tmp_buffer)));

      write(buffer.data(), buffer.size());
  }
```

Reply:
response looks fine, the result will no longer be 0.0 but 0

Reply:
Doesn't `std::stod` rely on locales? E.g. on some systems it may expect `,` to be used as floating point separator?

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/stof
Looks like there are no rely on locales.

Reply:
> I eventually came up with something like this:
> 
> ```
>     void operator()(const Number &number)
>     {
>         // we don't want to print NaN or Infinity
>         BOOST_ASSERT(std::isfinite(number.value));
>         // `fmt::memory_buffer` stores first 500 bytes in the object itself(i.e. on stack in this
>         // case) and then grows using heap if needed
>         fmt::memory_buffer buffer;
>         fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:f}"), number.value);
> 
>         // find out how many trailing zeros we have
>         size_t zeros = 0;
>         for (size_t index = buffer.size() - 1; index > 1; --index)
>         {
>             if (buffer[index] != '0' || buffer[index - 1] == '.') {
>                 break;
>             }
>             ++zeros;
>         }
>         // find position of the decimal point
>         size_t pointpos = std::find(buffer.begin(), buffer.end(), '.') - buffer.begin();
>         // we are removing trailing zeros and truncating the number to 10 digits after the decimal
>         buffer.resize(std::min(buffer.size() - zeros, pointpos + 10));
> 
>         write(buffer.data(), buffer.size());
>     }
> ```
> 
> Not sure there are no bugs so - will return to it later(or feel free to continue with it please :) )

use this will be fine
fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:.10f}"), number.value);

example:
fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:.f}"), 116.123456789); will returns 116.123456 
fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:.10f}"), 116.123456789); will returns 116.1234567890

Reply:
@SiarheiFedartsou I found a more reasonable way
```
void operator()(const Number &number)
{
    // we don't want to print NaN or Infinity
    BOOST_ASSERT(std::isfinite(number.value));
    // `fmt::memory_buffer` stores first 500 bytes in the object itself(i.e. on stack in this
    // case) and then grows using heap if needed
    fmt::memory_buffer buffer;
    fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:.10g}"), number.value);

    write(buffer.data(), buffer.size());
}
```
https://fmt.dev/latest/syntax.html#grammar-token-sf-precision
![Uploading image.png…]()